filebeat/input/filestream/input.go

```diff
@@ -177,12 +199,36 @@ func initState(log *logp.Logger, c loginp.Cursor, s fileSource) state {
 	return state
 }
 
-func (inp *filestream) open(log *logp.Logger, canceler input.Canceler, path string, offset int64) (reader.Reader, error) {
+func (inp *filestream) open(log *logp.Logger, canceler input.Canceler, srcID, path string, offset int64) (reader.Reader, error) {
 	f, err := inp.openFile(log, path, offset)
 	if err != nil {
 		return nil, err
 	}
 
+	if inp.detailedMetricsEnabled {
+		// if there was a previously running monitoring routine, stop it
+		if m, ok := inp.detailedMetrics[srcID]; ok {
+			m.cancel()
+		}
+
+		monitorCtx, monitorCancel := ctxtool.WithFunc(canceler, func() {
+			log.Debug("Stop monitoring of opened file")
+		})
+
+		inp.detailedMetrics[srcID] = newProgressMetrics(inp.ID, path, srcID, monitorCancel)
+		go func() {
+			timed.Periodic(monitorCtx, 30*time.Second, func() error {
+				fi, err := f.Stat()
+				if err != nil {
+					log.Errorf("failed to update current size of file for detailed metrics: %+v", err)
+					return nil
+				}
+				inp.detailedMetrics[srcID].updateCurrentSize(fi.Size())
+				return nil
+			})
+		}()
```

Comment:
alternatively to actively probing the file with `fstat`, we could define a function based metric to save CPU. The function based metric will query the size only on demand, and only if the last file size update is >10s.

Reply:
What do you mean by "only if the last file size update is >10s"? The file was updated 10 seconds ago?

Reply:
I mean, that instead of pro-actively updating the metrics over and over again, we could make it lazy via the metrics callback. Instead of spinning a go-routine we check the state on disk lazily.
The callback would cache the last size (or maybe even gets its update from the last file watcher scan), and only call fstat if the cached value has not been updated withing the last 10 seconds.

By making it 'cheaper' when the metrics are not required, we don't really need a setting to enable/disable metrics support. It will just always be there.